### PR TITLE
Add extended and correct description of normal ordering in policy evalua...

### PR DESCRIPTION
**Pull request by pasinskim**
...tion.

### CODE REVIEW

**Comment by pasinskim** [comment]
@nickanderson @estenberg how about such a description?

File: guide/language-concepts/normal-ordering.markdown [hunk 1]
```diff
@@ -19,19 +19,67 @@ You can override normal ordering in exceptional circumstances by making a
 promise in a class context and defining that class based on the outcome of
 another promise, or using the `depends_on` promise attribute.
 
-### Agent normal ordering
+## Agent normal ordering
 
 CFEngine tries to keep variable and class promises before starting to consider
-any other kind of promise. In this way, global variable and classes can be set.
+any other kind of promise. In this way, global variables and classes can be set.
```
Comment (nickanderson) [inline guide/language-concepts/normal-ordering.markdown]: can be set anywhere and used in later policy.

File: guide/language-concepts/normal-ordering.markdown [hunk 2]
```diff
@@ -19,19 +19,67 @@ You can override normal ordering in exceptional circumstances by making a
...
 
 If you set variables based on classes that are determined by other variables,
 then you introduce an order dependence to the resolution that might be
```
Comment (nickanderson) [inline guide/language-concepts/normal-ordering.markdown]: I don't understand "introduce an order dependance to the resolution that might be non-unique".

File: guide/language-concepts/normal-ordering.markdown [hunk 3]
```diff
@@ -19,19 +19,67 @@ You can override normal ordering in exceptional circumstances by making a
...
 non-unique. **Since CFEngine starts trying to converge values as soon as
 possible, it is best to define variables in bundles before using them**, i.e. as
-early as possible in your configuration. In general it is wise to avoid
-class-variable dependency as much as possible.
-
-CFEngine executes agent promise bundles in the strict order defined by the
+early as possible in your configuration. In order to make sure all global
```
Comment (nickanderson) [inline guide/language-concepts/normal-ordering.markdown]: What does "early as possible in your configuration" mean? Files listed in inputs first, bundles listed in bundlesequence first, both?

File: guide/language-concepts/normal-ordering.markdown [hunk 4]
```diff
@@ -19,19 +19,67 @@ You can override normal ordering in exceptional circumstances by making a
...
+variables and classes are available early enough policy pre-evaluation step was
+introduced.
+
+### Policy evaluation details
+
+Before exact evaluation of promises takes place first command line
+parameters are read and all classes defined using -D parameter are set. Next,
+environment detection is taking place and hard classes are discovered.
+When environment detection step is done all the persistent classes are 
+loaded and policy sanity check is performed using cf-promises.
```
Comment (nickanderson) [inline guide/language-concepts/normal-ordering.markdown]: A full or partial check? cf-promsies -cf ?

File: guide/language-concepts/normal-ordering.markdown [hunk 5]
```diff
@@ -19,19 +19,67 @@ You can override normal ordering in exceptional circumstances by making a
...
+bundles are resolved (in order in which files containing bundles were parsed) 
+and then variables in agent bundles.
+
+#### Agent evaluation step
+
+After pre-evaluation step is done normal evaluation is taking place.
+
+In this step CFEngine executes agent promise bundles in the strict order defined by the
 `bundlesequence` (possibly overridden by the `-b` or `--bundlesequence`
 command line option).  The `bundlesequence` defaults to just the `main`
```
Comment (nickanderson) [inline guide/language-concepts/normal-ordering.markdown]: No defaulting to the main bundle.

```
body common control
{
  inputs => { };
}

bundle agent main
{
  reports:
    "Hi from $(this.bundle)";
}
```
```
> $ cf-agent -KIf ./g.cf                                                       
2015-02-24T08:10:16-0600    error:  No bundlesequence in the common control body
2015-02-24T08:10:16-0600    error: Fatal CFEngine error: Errors in promise bundles: could not verify bundlesequence
2015-02-24T08:10:16-0600    error: Policy failed validation with command '"/home/nickanderson/.cfagent/bin/cf-promises" -c "./g.cf"'
```

File: guide/language-concepts/normal-ordering.markdown [hunk 6]
```diff
@@ -19,22 +19,69 @@ You can override normal ordering in exceptional circumstances by making a
...
+#### cf-promises policy validation step
+
+In this step policy is validated and classes and variables promises are evaluated. 
+Note that cached functions are executed here, and then again during the normal agent
+execution. Also policy validation is done using separate binary and any classes and
+variables are not preserved during validation step and normal agent run.
+
+#### Agent pre-evaluation step
+
+In order to support dynamic inputs and make sure all needed classes and
```
Comment (estenberg) [inline guide/language-concepts/normal-ordering.markdown]: I'm not sure if people know what "dynamic inputs" is. Perhaps say (i.e. expansion of variables in body common control inputs). In fact I think we should not say "dynamic inputs" at all - it's yet another concept that needs to be understood..

**Comment by estenberg** [comment]
Nice, there is a lot of details here.

In order to make this clearer, it would be great to start with a simple point-list of what happens, e.g.

1. -D option setting stuff
2. environment inventory discovery
3. persistent stuff..
4. cf-promises.
etc.

**Comment by pasinskim** [comment]
@estenberg I have added list summarizing the steps of policy evaluation. See the last commit.

**Comment by estenberg** [comment]
Looks great!

**Pull request merged**